File: retaRust/python_arch_reference/reta_architecture/persistence.py
from __future__ import annotations

"""SQLite persistence and audit layer for Reta architecture materialisations.

Persistence is intentionally orthogonal to the mathematical core.  It stores
instances, snapshots, cache materialisations and audit events; it does not define
what a topology, garbe, morphism, functor or natural transformation means.
"""

import hashlib
import json
import os
import sqlite3
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable, Mapping, Sequence


def _json_dumps(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), default=str)


def stable_digest(value: Any) -> str:
    return hashlib.sha256(_json_dumps(value).encode("utf-8")).hexdigest()


@dataclass(frozen=True)
class PersistenceConfig:
    db_path: str = ":memory:"
    initialise: bool = True
    journal_mode: str = "WAL"

    @classmethod
    def from_environment(cls, repo_root: Path | None = None) -> "PersistenceConfig":
        env_path = os.environ.get("RETA_PERSISTENCE_DB") or os.environ.get("RETA_AUDIT_DB")
        if env_path:
            return cls(db_path=env_path)
        return cls(db_path=":memory:")

    def snapshot(self) -> dict:
        return {"class": type(self).__name__, "db_path": self.db_path, "initialise": self.initialise, "journal_mode": self.journal_mode}


@dataclass(frozen=True)
class PersistedRecord:
    table: str
    key: str
    digest: str
    rowid: int | None = None

    def snapshot(self) -> dict:
        return {"class": type(self).__name__, "table": self.table, "key": self.key, "digest": self.digest, "rowid": self.rowid}


@dataclass(frozen=True)
class PersistenceBundle:
    config: PersistenceConfig

    def connect(self) -> sqlite3.Connection:
        if self.config.db_path != ":memory:":
            Path(self.config.db_path).expanduser().resolve().parent.mkdir(parents=True, exist_ok=True)
        connection = sqlite3.connect(self.config.db_path)
        connection.row_factory = sqlite3.Row
        if self.config.db_path != ":memory:" and self.config.journal_mode:
            try:
                connection.execute(f"PRAGMA journal_mode={self.config.journal_mode}")
            except sqlite3.DatabaseError:
                pass
        if self.config.initialise:
            initialise_persistence_schema(connection)
        return connection

    def snapshot(self) -> dict:
        return {
            "class": type(self).__name__,
            "category": "PersistenceCategory",
            "config": self.config.snapshot(),
            "tables": [
                "open_contexts",
                "local_sections",
                "sheaf_snapshots",
                "execution_runs",
                "audit_events",
                "cache_entries",
            ],
            "morphisms": [
                "persist_section",
                "persist_sections_batch",
                "load_section",
                "persist_sheaf_snapshot",
                "persist_sheaf_snapshots_batch",
                "persist_execution_run",
                "record_audit_event",
                "query_audit_events",
                "cache_put",
                "cache_put_many",
                "cache_get",
                "invalidate_cache",
            ],
            "universal_property": "load_persisted_snapshot_equals_original_snapshot_when_digest_matches",
        }


def bootstrap_persistence(repo_root: Path | None = None, config: PersistenceConfig | None = None, db_path: str | None = None) -> PersistenceBundle:
    if config is None:
        config = PersistenceConfig(db_path=db_path) if db_path is not None else PersistenceConfig.from_environment(repo_root)
    bundle = PersistenceBundle(config=config)
    if config.initialise:
        connection = bundle.connect()
        try:
            initialise_persistence_schema(connection)
        finally:
            connection.close()
    return bundle


def initialise_persistence_schema(connection: sqlite3.Connection) -> None:
    connection.executescript(
        """
        CREATE TABLE IF NOT EXISTS open_contexts (
            context_hash TEXT PRIMARY KEY,
            context_json TEXT NOT NULL,
            created_at REAL NOT NULL
        );
        CREATE TABLE IF NOT EXISTS local_sections (
            section_hash TEXT PRIMARY KEY,
            kind TEXT NOT NULL,
            name TEXT NOT NULL,
            context_hash TEXT,
            payload_json TEXT NOT NULL,
            created_at REAL NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_local_sections_kind_name ON local_sections(kind, name);
        CREATE TABLE IF NOT EXISTS sheaf_snapshots (
            snapshot_hash TEXT PRIMARY KEY,
            sheaf_name TEXT NOT NULL,
            context_hash TEXT,
            payload_json TEXT NOT NULL,
            created_at REAL NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_sheaf_snapshots_name ON sheaf_snapshots(sheaf_name);
        CREATE TABLE IF NOT EXISTS execution_runs (
            run_hash TEXT PRIMARY KEY,
            operation TEXT NOT NULL,
            context_hash TEXT,
            task_count INTEGER NOT NULL,
            payload_json TEXT NOT NULL,
            created_at REAL NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_execution_runs_operation ON execution_runs(operation);
        CREATE TABLE IF NOT EXISTS audit_events (
            event_id INTEGER PRIMARY KEY AUTOINCREMENT,
            event_type TEXT NOT NULL,
            subject TEXT NOT NULL,
            payload_hash TEXT NOT NULL,
            payload_json TEXT NOT NULL,
            created_at REAL NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_audit_events_type_subject ON audit_events(event_type, subject);
        CREATE TABLE IF NOT EXISTS cache_entries (
            cache_key TEXT PRIMARY KEY,
            value_hash TEXT NOT NULL,
            value_json TEXT NOT NULL,
            valid INTEGER NOT NULL DEFAULT 1,
            created_at REAL NOT NULL,
            updated_at REAL NOT NULL
        );
        """
    )
    connection.commit()


def persist_context(connection: sqlite3.Connection, context: Mapping[str, Any] | None) -> str | None:
    if context is None:
        return None
    payload = dict(context)
    digest = stable_digest(payload)
    connection.execute(
        "INSERT OR IGNORE INTO open_contexts(context_hash, context_json, created_at) VALUES (?, ?, ?)",
        (digest, _json_dumps(payload), time.time()),
    )
    return digest


def persist_section(
    connection: sqlite3.Connection,
    *,
    kind: str,
    name: str,
    payload: Any,
    context: Mapping[str, Any] | None = None,
) -> PersistedRecord:
    context_hash = persist_context(connection, context)
    section_payload = {"kind": kind, "name": name, "context_hash": context_hash, "payload": payload}
    digest = stable_digest(section_payload)
    connection.execute(
        """
        INSERT OR REPLACE INTO local_sections(section_hash, kind, name, context_hash, payload_json, created_at)
        VALUES (?, ?, ?, ?, ?, ?)
        """,
        (digest, kind, name, context_hash, _json_dumps(payload), time.time()),
    )
    connection.commit()
    return PersistedRecord("local_sections", f"{kind}:{name}", digest)


def load_section(connection: sqlite3.Connection, section_hash: str) -> dict | None:
    row = connection.execute("SELECT * FROM local_sections WHERE section_hash = ?", (section_hash,)).fetchone()
    if row is None:
        return None
    return {
        "section_hash": row["section_hash"],
        "kind": row["kind"],
        "name": row["name"],
        "context_hash": row["context_hash"],
        "payload": json.loads(row["payload_json"]),
        "created_at": row["created_at"],
    }


def persist_sheaf_snapshot(
    connection: sqlite3.Connection,
    *,
    sheaf_name: str,
    payload: Any,
    context: Mapping[str, Any] | None = None,
) -> PersistedRecord:
    context_hash = persist_context(connection, context)
    snapshot_payload = {"sheaf_name": sheaf_name, "context_hash": context_hash, "payload": payload}
    digest = stable_digest(snapshot_payload)
    connection.execute(
        """
        INSERT OR REPLACE INTO sheaf_snapshots(snapshot_hash, sheaf_name, context_hash, payload_json, created_at)
        VALUES (?, ?, ?, ?, ?)
        """,
        (digest, sheaf_name, context_hash, _json_dumps(payload), time.time()),
    )
    connection.commit()
    return PersistedRecord("sheaf_snapshots", sheaf_name, digest)


def load_sheaf_snapshot(connection: sqlite3.Connection, snapshot_hash: str) -> dict | None:
    row = connection.execute("SELECT * FROM sheaf_snapshots WHERE snapshot_hash = ?", (snapshot_hash,)).fetchone()
    if row is None:
        return None
    return {
        "snapshot_hash": row["snapshot_hash"],
        "sheaf_name": row["sheaf_name"],
        "context_hash": row["context_hash"],
        "payload": json.loads(row["payload_json"]),
        "created_at": row["created_at"],
    }


def persist_execution_run(
    connection: sqlite3.Connection,
    *,
    operation: str,
    task_count: int,
    payload: Any,
    context: Mapping[str, Any] | None = None,
) -> PersistedRecord:
    context_hash = persist_context(connection, context)
    run_payload = {"operation": operation, "task_count": int(task_count), "context_hash": context_hash, "payload": payload}
    digest = stable_digest(run_payload)
    connection.execute(
        """
        INSERT OR REPLACE INTO execution_runs(run_hash, operation, context_hash, task_count, payload_json, created_at)
        VALUES (?, ?, ?, ?, ?, ?)
        """,
        (digest, operation, context_hash, int(task_count), _json_dumps(payload), time.time()),
    )
    connection.commit()
    return PersistedRecord("execution_runs", operation, digest)


def record_audit_event(connection: sqlite3.Connection, *, event_type: str, subject: str, payload: Any) -> PersistedRecord:
    digest = stable_digest(payload)
    cursor = connection.execute(
        """
        INSERT INTO audit_events(event_type, subject, payload_hash, payload_json, created_at)
        VALUES (?, ?, ?, ?, ?)
        """,
        (event_type, subject, digest, _json_dumps(payload), time.time()),
    )
    connection.commit()
    return PersistedRecord("audit_events", f"{event_type}:{subject}", digest, rowid=cursor.lastrowid)


def query_audit_events(connection: sqlite3.Connection, *, event_type: str | None = None, subject: str | None = None, limit: int = 50) -> list[dict]:
    clauses = []
    values: list[Any] = []
    if event_type is not None:
        clauses.append("event_type = ?")
        values.append(event_type)
    if subject is not None:
        clauses.append("subject = ?")
        values.append(subject)
    where = " WHERE " + " AND ".join(clauses) if clauses else ""
    values.append(max(1, int(limit)))
    rows = connection.execute(
        f"SELECT * FROM audit_events{where} ORDER BY event_id DESC LIMIT ?",
        tuple(values),
    ).fetchall()
    return [
        {
            "event_id": row["event_id"],
            "event_type": row["event_type"],
            "subject": row["subject"],
            "payload_hash": row["payload_hash"],
            "payload": json.loads(row["payload_json"]),
            "created_at": row["created_at"],
        }
        for row in rows
    ]


def cache_put(connection: sqlite3.Connection, cache_key: str, value: Any) -> PersistedRecord:
    digest = stable_digest(value)
    now = time.time()
    connection.execute(
        """
        INSERT INTO cache_entries(cache_key, value_hash, value_json, valid, created_at, updated_at)
        VALUES (?, ?, ?, 1, ?, ?)
        ON CONFLICT(cache_key) DO UPDATE SET
            value_hash = excluded.value_hash,
            value_json = excluded.value_json,
            valid = 1,
            updated_at = excluded.updated_at
        """,
        (cache_key, digest, _json_dumps(value), now, now),
    )
    connection.commit()
    return PersistedRecord("cache_entries", cache_key, digest)


def cache_get(connection: sqlite3.Connection, cache_key: str) -> Any | None:
    row = connection.execute("SELECT value_json FROM cache_entries WHERE cache_key = ? AND valid = 1", (cache_key,)).fetchone()
    if row is None:
        return None
    return json.loads(row["value_json"])


def invalidate_cache(connection: sqlite3.Connection, cache_key: str | None = None) -> int:
    if cache_key is None:
        cursor = connection.execute("UPDATE cache_entries SET valid = 0, updated_at = ? WHERE valid = 1", (time.time(),))
    else:
        cursor = connection.execute("UPDATE cache_entries SET valid = 0, updated_at = ? WHERE cache_key = ?", (time.time(), cache_key))
    connection.commit()
    return int(cursor.rowcount)


def _prepare_cache_entries_worker(payload):
    entries = payload
    prepared = []
    for cache_key, value in entries:
        prepared.append((str(cache_key), stable_digest(value), _json_dumps(value)))
    return prepared


def _prepare_section_entries_worker(payload):
    entries = payload
    prepared = []
    for kind, name, value_payload, context in entries:
        context_hash = stable_digest(dict(context)) if context is not None else None
        section_payload = {"kind": kind, "name": name, "context_hash": context_hash, "payload": value_payload}
        prepared.append((kind, name, context_hash, _json_dumps(value_payload), stable_digest(section_payload), dict(context) if context is not None else None))
    return prepared


def _prepare_sheaf_snapshot_entries_worker(payload):
    entries = payload
    prepared = []
    for sheaf_name, value_payload, context in entries:
        context_hash = stable_digest(dict(context)) if context is not None else None
        snapshot_payload = {"sheaf_name": sheaf_name, "context_hash": context_hash, "payload": value_payload}
        prepared.append((sheaf_name, context_hash, _json_dumps(value_payload), stable_digest(snapshot_payload), dict(context) if context is not None else None))
    return prepared


def _prepare_persistence_entries_in_processes(entries: Sequence, worker, operation: str):
    try:
        from .parallel_execution import ParallelExecutionConfig, _chunks, _pool_map_ordered

        config = ParallelExecutionConfig.from_environment()
        if not config.should_use_processes(len(entries)):
            return None
        chunks = list(_chunks(list(entries), config.chunk_size))
        if len(chunks) <= 1:
            return None
        chunk_results, _workers = _pool_map_ordered(worker, chunks, config)
        prepared = []
        for chunk in chunk_results:
            prepared.extend(chunk)
        return prepared
    except Exception:
        return None


def cache_put_many(connection: sqlite3.Connection, entries: Sequence[tuple[str, Any]]) -> list[PersistedRecord]:
    """Store many cache values, computing JSON/digests in process chunks when useful.

    SQLite writes remain serial in this process.  Only pure payload preparation is
    parallelised, so database locking and cache semantics stay deterministic.
    """
    prepared = _prepare_persistence_entries_in_processes(entries, _prepare_cache_entries_worker, "cache_put_many")
    if prepared is None:
        prepared = []
        for cache_key, value in entries:
            prepared.append((str(cache_key), stable_digest(value), _json_dumps(value)))
    now = time.time()
    records: list[PersistedRecord] = []
    for cache_key, digest, value_json in prepared:
        connection.execute(
            """
            INSERT INTO cache_entries(cache_key, value_hash, value_json, valid, created_at, updated_at)
            VALUES (?, ?, ?, 1, ?, ?)
            ON CONFLICT(cache_key) DO UPDATE SET
                value_hash = excluded.value_hash,
                value_json = excluded.value_json,
                valid = 1,
                updated_at = excluded.updated_at
            """,
            (cache_key, digest, value_json, now, now),
        )
        records.append(PersistedRecord("cache_entries", cache_key, digest))
    connection.commit()
    return records


def persist_sections_batch(
    connection: sqlite3.Connection,
    entries: Sequence[tuple[str, str, Any, Mapping[str, Any] | None]],
) -> list[PersistedRecord]:
    """Persist many local sections with process-parallel digest preparation."""
    prepared = _prepare_persistence_entries_in_processes(entries, _prepare_section_entries_worker, "persist_sections_batch")
    if prepared is None:
        prepared = _prepare_section_entries_worker(entries)
    now = time.time()
    records: list[PersistedRecord] = []
    for kind, name, context_hash, payload_json, digest, context in prepared:
        if context is not None:
            connection.execute(
                "INSERT OR IGNORE INTO open_contexts(context_hash, context_json, created_at) VALUES (?, ?, ?)",
                (context_hash, _json_dumps(context), now),
            )
        connection.execute(
            """
            INSERT OR REPLACE INTO local_sections(section_hash, kind, name, context_hash, payload_json, created_at)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (digest, kind, name, context_hash, payload_json, now),
        )
        records.append(PersistedRecord("local_sections", f"{kind}:{name}", digest))
    connection.commit()
    return records


def persist_sheaf_snapshots_batch(
    connection: sqlite3.Connection,
    entries: Sequence[tuple[str, Any, Mapping[str, Any] | None]],
) -> list[PersistedRecord]:
    """Persist many sheaf snapshots with process-parallel digest preparation."""
    prepared = _prepare_persistence_entries_in_processes(entries, _prepare_sheaf_snapshot_entries_worker, "persist_sheaf_snapshots_batch")
    if prepared is None:
        prepared = _prepare_sheaf_snapshot_entries_worker(entries)
    now = time.time()
    records: list[PersistedRecord] = []
    for sheaf_name, context_hash, payload_json, digest, context in prepared:
        if context is not None:
            connection.execute(
                "INSERT OR IGNORE INTO open_contexts(context_hash, context_json, created_at) VALUES (?, ?, ?)",
                (context_hash, _json_dumps(context), now),
            )
        connection.execute(
            """
            INSERT OR REPLACE INTO sheaf_snapshots(snapshot_hash, sheaf_name, context_hash, payload_json, created_at)
            VALUES (?, ?, ?, ?, ?)
            """,
            (digest, sheaf_name, context_hash, payload_json, now),
        )
        records.append(PersistedRecord("sheaf_snapshots", sheaf_name, digest))
    connection.commit()
    return records

File: retaRust/crates/reta_architecture/src/persistence.rs
//! Persistence and audit morphisms transcompiled from
//! `python_arch_reference/reta_architecture/persistence.py`.
//!
//! The Python source owns SQLite materialisation.  This Rust stage keeps the
//! same table names, record shapes, cache/audit operations and universal
//! property in a dependency-free in-memory store.  A later stage can swap the
//! store implementation for SQLite without changing callers.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

pub fn stable_digest_text(value: &str) -> String {
    // Dependency-free stable FNV-1a digest.  The Python layer uses SHA-256;
    // this is a deterministic architecture key until the concrete DB backend
    // gets wired with a crypto crate or the standard library gains one.
    let mut hash: u64 = 0xcbf29ce484222325;
    for byte in value.as_bytes() {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(0x100000001b3);
    }
    format!("{hash:016x}")
}

fn jsonish_pair_map(items: &BTreeMap<String, String>) -> String {
    let mut text = String::from("{");
    for (idx, (key, value)) in items.iter().enumerate() {
        if idx > 0 {
            text.push(',');
        }
        text.push_str(&format!("{:?}:{:?}", key, value));
    }
    text.push('}');
    text
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistenceConfig {
    pub db_path: String,
    pub initialise: bool,
    pub journal_mode: String,
}

impl Default for PersistenceConfig {
    fn default() -> Self {
        Self {
            db_path: ":memory:".to_string(),
            initialise: true,
            journal_mode: "WAL".to_string(),
        }
    }
}

impl PersistenceConfig {
    pub fn from_environment() -> Self {
        let db_path = std::env::var("RETA_PERSISTENCE_DB")
            .ok()
            .or_else(|| std::env::var("RETA_AUDIT_DB").ok())
            .unwrap_or_else(|| ":memory:".to_string());
        Self {
            db_path,
            ..Self::default()
        }
    }

    pub fn snapshot(&self) -> PersistenceConfigSnapshot {
        PersistenceConfigSnapshot {
            class: "PersistenceConfig".to_string(),
            db_path: self.db_path.clone(),
            initialise: self.initialise,
            journal_mode: self.journal_mode.clone(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistenceConfigSnapshot {
    pub class: String,
    pub db_path: String,
    pub initialise: bool,
    pub journal_mode: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistedRecord {
    pub table: String,
    pub key: String,
    pub digest: String,
    pub rowid: Option<u64>,
}

impl PersistedRecord {
    pub fn new(
        table: impl Into<String>,
        key: impl Into<String>,
        digest: impl Into<String>,
    ) -> Self {
        Self {
            table: table.into(),
            key: key.into(),
            digest: digest.into(),
            rowid: None,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistedSection {
    pub section_hash: String,
    pub kind: String,
    pub name: String,
    pub context_hash: Option<String>,
    pub payload_json: String,
    pub created_at_counter: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct SheafSnapshotRecord {
    pub snapshot_hash: String,
    pub sheaf_name: String,
    pub context_hash: Option<String>,
    pub payload_json: String,
    pub created_at_counter: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ExecutionRunRecord {
    pub run_hash: String,
    pub operation: String,
    pub context_hash: Option<String>,
    pub task_count: usize,
    pub payload_json: String,
    pub created_at_counter: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct AuditEventRecord {
    pub event_id: u64,
    pub event_type: String,
    pub subject: String,
    pub payload_hash: String,
    pub payload_json: String,
    pub created_at_counter: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct CacheEntryRecord {
    pub cache_key: String,
    pub value_hash: String,
    pub value_json: String,
    pub valid: bool,
    pub created_at_counter: u64,
    pub updated_at_counter: u64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistenceSnapshot {
    pub class: String,
    pub category: String,
    pub config: PersistenceConfigSnapshot,
    pub tables: Vec<String>,
    pub morphisms: Vec<String>,
    pub universal_property: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistenceStore {
    pub contexts: BTreeMap<String, String>,
    pub local_sections: BTreeMap<String, PersistedSection>,
    pub sheaf_snapshots: BTreeMap<String, SheafSnapshotRecord>,
    pub execution_runs: BTreeMap<String, ExecutionRunRecord>,
    pub audit_events: Vec<AuditEventRecord>,
    pub cache_entries: BTreeMap<String, CacheEntryRecord>,
    counter: u64,
}

impl Default for PersistenceStore {
    fn default() -> Self {
        Self {
            contexts: BTreeMap::new(),
            local_sections: BTreeMap::new(),
            sheaf_snapshots: BTreeMap::new(),
            execution_runs: BTreeMap::new(),
            audit_events: Vec::new(),
            cache_entries: BTreeMap::new(),
            counter: 0,
        }
    }
}

impl PersistenceStore {
    fn tick(&mut self) -> u64 {
        self.counter += 1;
        self.counter
    }

    pub fn persist_context(&mut self, context: Option<BTreeMap<String, String>>) -> Option<String> {
        let payload = context?;
        let text = jsonish_pair_map(&payload);
        let digest = stable_digest_text(&text);
        self.contexts.entry(digest.clone()).or_insert(text);
        Some(digest)
    }

    pub fn persist_section(
        &mut self,
        kind: impl Into<String>,
        name: impl Into<String>,
        payload_json: impl Into<String>,
        context: Option<BTreeMap<String, String>>,
    ) -> PersistedRecord {
        let kind = kind.into();
        let name = name.into();
        let payload_json = payload_json.into();
        let context_hash = self.persist_context(context);
        let digest = stable_digest_text(&format!(
            "kind={kind};name={name};context={:?};payload={payload_json}",
            context_hash
        ));
        let created = self.tick();
        self.local_sections.insert(
            digest.clone(),
            PersistedSection {
                section_hash: digest.clone(),
                kind: kind.clone(),
                name: name.clone(),
                context_hash,
                payload_json,
                created_at_counter: created,
            },
        );
        PersistedRecord::new("local_sections", format!("{kind}:{name}"), digest)
    }

    pub fn load_section(&self, section_hash: &str) -> Option<&PersistedSection> {
        self.local_sections.get(section_hash)
    }

    pub fn persist_sheaf_snapshot(
        &mut self,
        sheaf_name: impl Into<String>,
        payload_json: impl Into<String>,
        context: Option<BTreeMap<String, String>>,
    ) -> PersistedRecord {
        let sheaf_name = sheaf_name.into();
        let payload_json = payload_json.into();
        let context_hash = self.persist_context(context);
        let digest = stable_digest_text(&format!(
            "sheaf={sheaf_name};context={:?};payload={payload_json}",
            context_hash
        ));
        let created = self.tick();
        self.sheaf_snapshots.insert(
            digest.clone(),
            SheafSnapshotRecord {
                snapshot_hash: digest.clone(),
                sheaf_name: sheaf_name.clone(),
                context_hash,
                payload_json,
                created_at_counter: created,
            },
        );
        PersistedRecord::new("sheaf_snapshots", sheaf_name, digest)
    }

    pub fn persist_execution_run(
        &mut self,
        operation: impl Into<String>,
        task_count: usize,
        payload_json: impl Into<String>,
        context: Option<BTreeMap<String, String>>,
    ) -> PersistedRecord {
        let operation = operation.into();
        let payload_json = payload_json.into();
        let context_hash = self.persist_context(context);
        let digest = stable_digest_text(&format!(
            "operation={operation};tasks={task_count};context={:?};payload={payload_json}",
            context_hash
        ));
        let created = self.tick();
        self.execution_runs.insert(
            digest.clone(),
            ExecutionRunRecord {
                run_hash: digest.clone(),
                operation: operation.clone(),
                context_hash,
                task_count,
                payload_json,
                created_at_counter: created,
            },
        );
        PersistedRecord::new("execution_runs", operation, digest)
    }

    pub fn record_audit_event(
        &mut self,
        event_type: impl Into<String>,
        subject: impl Into<String>,
        payload_json: impl Into<String>,
    ) -> PersistedRecord {
        let event_type = event_type.into();
        let subject = subject.into();
        let payload_json = payload_json.into();
        let digest = stable_digest_text(&payload_json);
        let created = self.tick();
        let event_id = self.audit_events.len() as u64 + 1;
        self.audit_events.push(AuditEventRecord {
            event_id,
            event_type: event_type.clone(),
            subject: subject.clone(),
            payload_hash: digest.clone(),
            payload_json,
            created_at_counter: created,
        });
        PersistedRecord {
            table: "audit_events".to_string(),
            key: format!("{event_type}:{subject}"),
            digest,
            rowid: Some(event_id),
        }
    }

    pub fn query_audit_events(
        &self,
        event_type: Option<&str>,
        subject: Option<&str>,
        limit: usize,
    ) -> Vec<AuditEventRecord> {
        let mut out = self
            .audit_events
            .iter()
            .rev()
            .filter(|event| {
                event_type
                    .map(|wanted| wanted == event.event_type)
                    .unwrap_or(true)
            })
            .filter(|event| {
                subject
                    .map(|wanted| wanted == event.subject)
                    .unwrap_or(true)
            })
            .take(limit.max(1))
            .cloned()
            .collect::<Vec<_>>();
        out.sort_by(|left, right| right.event_id.cmp(&left.event_id));
        out
    }

    pub fn cache_put(
        &mut self,
        cache_key: impl Into<String>,
        value_json: impl Into<String>,
    ) -> PersistedRecord {
        let cache_key = cache_key.into();
        let value_json = value_json.into();
        let digest = stable_digest_text(&value_json);
        let now = self.tick();
        let created = self
            .cache_entries
            .get(&cache_key)
            .map(|entry| entry.created_at_counter)
            .unwrap_or(now);
        self.cache_entries.insert(
            cache_key.clone(),
            CacheEntryRecord {
                cache_key: cache_key.clone(),
                value_hash: digest.clone(),
                value_json,
                valid: true,
                created_at_counter: created,
                updated_at_counter: now,
            },
        );
        PersistedRecord::new("cache_entries", cache_key, digest)
    }

    pub fn cache_put_many(&mut self, entries: &[(String, String)]) -> Vec<PersistedRecord> {
        entries
            .iter()
            .map(|(key, value)| self.cache_put(key.clone(), value.clone()))
            .collect()
    }

    pub fn cache_get(&self, cache_key: &str) -> Option<String> {
        self.cache_entries
            .get(cache_key)
            .filter(|entry| entry.valid)
            .map(|entry| entry.value_json.clone())
    }

    pub fn invalidate_cache(&mut self, cache_key: Option<&str>) -> usize {
        let now = self.tick();
        let mut changed = 0usize;
        for (key, entry) in self.cache_entries.iter_mut() {
            if cache_key.map(|wanted| wanted == key).unwrap_or(true) && entry.valid {
                entry.valid = false;
                entry.updated_at_counter = now;
                changed += 1;
            }
        }
        changed
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PersistenceBundle {
    pub config: PersistenceConfig,
}

impl PersistenceBundle {
    pub fn new(config: PersistenceConfig) -> Self {
        Self { config }
    }

    pub fn in_memory_store(&self) -> PersistenceStore {
        let _ = self.config.initialise;
        PersistenceStore::default()
    }

    pub fn snapshot(&self) -> PersistenceSnapshot {
        PersistenceSnapshot {
            class: "PersistenceBundle".to_string(),
            category: "PersistenceCategory".to_string(),
            config: self.config.snapshot(),
            tables: vec![
                "open_contexts".to_string(),
                "local_sections".to_string(),
                "sheaf_snapshots".to_string(),
                "execution_runs".to_string(),
                "audit_events".to_string(),
                "cache_entries".to_string(),
                "activation_stores_via_local_sections".to_string(),
            ],
            morphisms: vec![
                "persist_section".to_string(),
                "persist_sections_batch".to_string(),
                "load_section".to_string(),
                "persist_sheaf_snapshot".to_string(),
                "persist_sheaf_snapshots_batch".to_string(),
                "persist_execution_run".to_string(),
                "record_audit_event".to_string(),
                "query_audit_events".to_string(),
                "cache_put".to_string(),
                "cache_put_many".to_string(),
                "cache_get".to_string(),
                "invalidate_cache".to_string(),
                "persist_activation_store_text".to_string(),
                "load_activation_store_text".to_string(),
                "audit_activation_store_persistence".to_string(),
                "cache_latest_activation_store_digest".to_string(),
            ],
            universal_property:
                "load_persisted_snapshot_equals_original_snapshot_when_digest_matches".to_string(),
        }
    }
}

pub fn bootstrap_persistence(
    config: Option<PersistenceConfig>,
    db_path: Option<String>,
) -> PersistenceBundle {
    let config = config.unwrap_or_else(|| {
        db_path
            .map(|path| PersistenceConfig {
                db_path: path,
                ..PersistenceConfig::default()
            })
            .unwrap_or_else(PersistenceConfig::from_environment)
    });
    PersistenceBundle::new(config)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cache_roundtrip_and_invalidate() {
        let bundle = bootstrap_persistence(Some(PersistenceConfig::default()), None);
        let mut store = bundle.in_memory_store();
        let record = store.cache_put("a", "{\"x\":1}");
        assert_eq!(record.table, "cache_entries");
        assert_eq!(store.cache_get("a"), Some("{\"x\":1}".to_string()));
        assert_eq!(store.invalidate_cache(Some("a")), 1);
        assert_eq!(store.cache_get("a"), None);
    }

    #[test]
    fn audit_query_is_newest_first() {
        let mut store = PersistenceStore::default();
        store.record_audit_event("build", "a", "1");
        store.record_audit_event("build", "b", "2");
        let events = store.query_audit_events(Some("build"), None, 10);
        assert_eq!(events[0].subject, "b");
    }
}

// Stage 16: concrete persistence.py compatibility wrappers.
pub fn stable_digest(value: &str) -> String { stable_digest_text(value) }

pub fn _json_dumps(items: &BTreeMap<String, String>) -> String { jsonish_pair_map(items) }

pub fn connect(db_path: Option<String>) -> PersistenceBundle {
    bootstrap_persistence(None, db_path)
}

pub fn initialise_persistence_schema(bundle: &PersistenceBundle) -> PersistenceSnapshot {
    bundle.snapshot()
}

pub fn _prepare_section_entries_worker(items: &[(String, String, String)]) -> Vec<(String, String, String, String)> {
    items.iter().map(|(kind, name, payload)| (kind.clone(), name.clone(), payload.clone(), stable_digest_text(payload))).collect()
}

pub fn _prepare_sheaf_snapshot_entries_worker(items: &[(String, String)]) -> Vec<(String, String, String)> {
    items.iter().map(|(name, payload)| (name.clone(), payload.clone(), stable_digest_text(payload))).collect()
}

pub fn _prepare_cache_entries_worker(items: &[(String, String)]) -> Vec<(String, String, String)> {
    items.iter().map(|(key, value)| (key.clone(), value.clone(), stable_digest_text(value))).collect()
}

pub fn _prepare_persistence_entries_in_processes(items: &[(String, String)]) -> Vec<(String, String, String)> {
    _prepare_cache_entries_worker(items)
}

pub fn persist_sections_batch(store: &mut PersistenceStore, sections: &[(String, String, String)]) -> Vec<PersistedRecord> {
    sections.iter().map(|(kind, name, payload)| store.persist_section(kind.clone(), name.clone(), payload.clone(), None)).collect()
}

pub fn persist_sheaf_snapshots_batch(store: &mut PersistenceStore, snapshots: &[(String, String)]) -> Vec<PersistedRecord> {
    snapshots.iter().map(|(name, payload)| store.persist_sheaf_snapshot(name.clone(), payload.clone(), None)).collect()
}

pub fn load_sheaf_snapshot(store: &PersistenceStore, snapshot_hash: &str) -> Option<SheafSnapshotRecord> {
    store.sheaf_snapshots.get(snapshot_hash).cloned()
}

// Stage 15: explicit py-reta-arch compatibility surface markers.
// These markers keep historical Python architecture symbol names visible
// while the Rust implementation is migrated module by module. They are
// not a claim of byte-exact semantic replacement for every listed helper.
#[allow(dead_code)]
pub const PY_ARCH_STAGE15_SURFACE: &[&str] = &[
    "_json_dumps",
    "_prepare_cache_entries_worker",
    "_prepare_persistence_entries_in_processes",
    "_prepare_section_entries_worker",
    "_prepare_sheaf_snapshot_entries_worker",
    "connect",
    "initialise_persistence_schema",
    "load_sheaf_snapshot",
    "stable_digest",
    "persist_sections_batch",
    "persist_sheaf_snapshots_batch",
];

#[allow(dead_code)]
pub fn stage15_py_surface_names() -> &'static [&'static str] {
    PY_ARCH_STAGE15_SURFACE
}
